homeassistant/components/sensor/sht31.py

```diff
@@ -0,0 +1,111 @@
+"""
+Support for Sensirion SHT31 temperature and humidity sensor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.sht31/
+"""
+
+import logging
+import math
+import voluptuous as vol
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import (
+    TEMP_FAHRENHEIT, CONF_NAME, CONF_MONITORED_CONDITIONS)
+from homeassistant.helpers.entity import Entity
+from homeassistant.util.temperature import celsius_to_fahrenheit
+
+REQUIREMENTS = ['Adafruit-GPIO==1.0.3',
+                'Adafruit-SHT31==1.0.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_I2C_ADDRESS = 'i2c_address'
+
+DEFAULT_NAME = 'SHT31'
+DEFAULT_I2C_ADDRESS = '0x44'
+
+SENSOR_TEMPERATURE = 'temperature'
+SENSOR_HUMIDITY = 'humidity'
+SENSOR_TYPES = {
+    SENSOR_TEMPERATURE: ['Temperature', None],
+    SENSOR_HUMIDITY: ['Humidity', '%']
+}
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_I2C_ADDRESS, default=DEFAULT_I2C_ADDRESS): cv.string,
```

Comment:
I don't know how much you know about YAML (the config file format), but using hexadecimal options in YAML can have unexpected consequences. For example, if you type

```yaml
abc: 0xFF
```

in a YAML file, the parser will automatically convert this into an integer for you (according to [YAML spec](http://yaml.org/spec/1.0/)), so your Python code will receive `abc` as being `255`, thus breaking i2c.

So users will have to explicitly put quotation marks around the i2c address:

```yaml
abc: '0xFF'
```

(I know [BS1750](https://home-assistant.io/components/sensor.bh1750/) does the same, but a comment in the docs would be nice)

Reply:
Thanks for the thorough review. I was not aware of this automatic conversion and the result was unexpected indeed. I've changed the documentation to use a quoted address and added better error handling in the code.

On the other hand, the way BS1750 works (without quotes, auto conversion) doesn't look wrong either. In the end the I2C library is happy with int 255 also, maybe I just need to convert it back to hex for the user in case of error. I'm not sure I understand why you recommend against using unquoted addresses in then config. I think it looks more straightforward for the user and easy to handle in the code.

Reply:
Yes, that would of course also be an option and there's nothing wrong with it. The problem is that your current code would blow up with unquoted addresses. In this code block:

```python
i2c_address = int(config.get(CONF_I2C_ADDRESS), 16)
```

with an unquoted address (like `i2c_address: 0xFF`), `config.get(CONF_I2C_ADDRESS)` will return the string `"255"`. But `int("255", 16)` will think you're trying to say `0x255` and so `i2c_address` is now `597` even though we wanted 255/0xFF.


<img width="165" alt="screen shot 2018-03-10 at 14 46 35" src="https://user-images.githubusercontent.com/6833237/37242953-e93e68ea-2471-11e8-9b0e-7caa610e0e3a.png">

Reply:
Thanks for pointing out the double conversion. I'll correct that and go for the option without quotes.

Reply:
Ok, that's great and will also help with invalid config validation **before** runtime.

Just a minor issue though: in your platform schema, you're still saying that the address should be a string:

```python
vol.Optional(CONF_I2C_ADDRESS, default=DEFAULT_I2C_ADDRESS): cv.string,
```

voluptuous can however check that it is an *integer* before runtime and can help spot mistakes before starting Home Assistant. So to fully "coerce" the i2c_address to be an integer, you can do the following:

```python
vol.Optional(CONF_I2C_ADDRESS, default=DEFAULT_I2C_ADDRESS): cv.positive_int,
```

Looking at [the datasheet](http://www.mouser.com/ds/2/682/Sensirion_Humidity_Sensors_SHT3x_Datasheet_digital-971521.pdf), only two i2c addresses are supported anyway. So we could even further improve the config validation with the following:

```python
vol.Optional(CONF_I2C_ADDRESS, default=DEFAULT_I2C_ADDRESS): 
    vol.All(vol.Coerce(int), vol.Range(min=0x44, max=0x45))
```

(Then you could to change `i2c_address = int(config.get(CONF_I2C_ADDRESS))` to `i2c_address = config.get(CONF_I2C_ADDRESS)`)